av_metrics/src/capi.rs

```diff
@@ -0,0 +1,381 @@
+                .expect(&("Failed to decode the ".to_owned() + metric + " y4m file")),
+        }
+    }
+}
+
+#[inline(always)]
+fn convert_c_string_into_path(c_buf: *const c_char) -> PathBuf {
+    let c_str = unsafe { CStr::from_ptr(c_buf) };
+    Path::new(c_str.to_str().unwrap()).to_path_buf()
+}
+
+#[inline(always)]
+fn run_metric(
+    path1: *const c_char,
+    path2: *const c_char,
+    frame_limit: usize,
+    metric: &str,
+    is_frame: bool,
+) -> (*const MetricContext, f64) {
+    if path1.is_null() || path2.is_null() {
+        return (null(), -1.0);
+    }
+
+    let path1 = convert_c_string_into_path(path1);
+    let path2 = convert_c_string_into_path(path2);
+
+    let input_type1 = InputType::detect(&path1);
+    let input_type2 = InputType::detect(&path2);
+
+    match (input_type1, input_type2) {
+        (InputType::Video(c1), InputType::Video(c2)) => {
+            if is_frame {
+                return run_frame_metric(path1, c1, path2, c2, frame_limit, metric);
+            } else {
+                return run_video_metric(path1, c1, path2, c2, frame_limit, metric);
+            }
+        }
+        (InputType::Unknown, _) | (_, InputType::Unknown) => {}
+    }
+
+    (null(), -1.0)
+}
+
+#[inline(always)]
+fn run_video_metric<P: AsRef<Path>>(
+    path1: P,
+    container1: VideoContainer,
+    path2: P,
+    container2: VideoContainer,
+    frame_limit: usize,
+    metric: &str,
+) -> (*const MetricContext, f64) {
+    let mut file1 =
+        File::open(path1).expect(&("Error opening the first ".to_owned() + metric + " video"));
+    let mut file2 =
+        File::open(path2).expect(&("Error opening the second ".to_owned() + metric + " video"));
+
+    let mut dec1 = container1.get_decoder(&mut file1, &("first".to_owned() + metric));
+    let mut dec2 = container2.get_decoder(&mut file2, &("second".to_owned() + metric));
+
+    let mut limit: Option<usize> = None;
+    if frame_limit > 0 {
+        limit = Some(frame_limit);
+    }
+    if metric == "ciede" {
+        let val = ciede::calculate_video_ciede(&mut dec1, &mut dec2, limit);
+        if let Ok(metric) = val {
+            return (null(), metric);
+        }
+    }
+    let val = match metric {
+        "psnr" => psnr::calculate_video_psnr(&mut dec1, &mut dec2, limit),
+        "apsnr" => psnr::calculate_video_apsnr(&mut dec1, &mut dec2, limit),
+        "psnr_hvs" => psnr_hvs::calculate_video_psnr_hvs(&mut dec1, &mut dec2, limit),
+        "ssim" => ssim::calculate_video_ssim(&mut dec1, &mut dec2, limit),
+        _ => ssim::calculate_video_msssim(&mut dec1, &mut dec2, limit),
+    };
+    if let Ok(metric) = val {
+        let ctx = MetricContext {
+            y: metric.y,
+            u: metric.u,
+            v: metric.v,
+            avg: metric.avg,
+        };
+        let boxed = Box::new(ctx);
+        return (Box::into_raw(boxed), 0.0);
+    }
+    (null(), -1.0)
+}
+
+#[inline(always)]
+fn run_frame_metric<P: AsRef<Path>>(
```

Comment:
Might be a good idea to provide a mean to run the metric on frames and not only files.

Reply:
Yep, I thought about it, but I decided to discard that idea because I didn't know how to implement the specific. I mean, should we implement frame API that read from two images or from specific buffers?

Reply:
@lu-zero 
Do you think something like this could work? 

```
const *MetricType calculate_frame_metric(
    const uint8_t *frame1_data[3] ,
    const uint8_t *frame2_data[3],
    uint32_t width,
    uint32_t height,
    enum Colorspace color
);
```

Reply:
I'd add 2 other args with `usize strides[3]` to be fully generic

Reply:
Sure, thanks! :)

Oh, if the bitdepth is higher? Should I create a new function or a union that wraps the two different data types?

Reply:
just add a field for the bitdepth, the conversion may happen internally.

Reply:
But if the data are originally 16-bit for example? This type of function wouldn't be pretty comfortable as-is. 
```
const *MetricType calculate_frame_metric(
    const uint8_t *frame1_data[3],
    const usize frame1_strides[3],
    const uint8_t *frame2_data[3],
    const usize frame2_strides[3],
    uint32_t width,
    uint32_t height,
    uint32_t bitdepth,
    enum Colorspace color
)
```

Reply:
why not? as long the pointers are aligned correctly it is fine.

Reply:
Yeah, just trying to further simplify the user's life. Having two types of functions would be better for a user, don't know, but let's leave it the way it is :)